beacon_node/beacon_chain/src/beacon_block_reward.rs

```diff
@@ -169,19 +169,75 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
-            .attestation_rewards
-            .total;
-
-        Ok(block_attestation_reward)
+        // In phase0, rewards for including attestations are awarded at epoch boundaries when the corresponding
+        // attestations are contained in state.previous_epoch_attestations. So, if an attestation within this block has
+        // target = previous_epoch, it is directly inserted into previous_epoch_attestations and we need the state at
+        // the end of this epoch, or the attestation has target = current_epoch and thus we need the state at the end
+        // of the next epoch.
+        // We fetch these lazily, as only one might be needed depending on the block's content.
+        let mut current_epoch_end = None;
+        let mut next_epoch_end = None;
+
+        let epoch = block.epoch();
+        let mut block_reward = 0;
+
+        for attestation in block.body().attestations() {
+            let processing_epoch_end = if attestation.data().target.epoch == epoch {
+                let next_epoch_end = match &mut next_epoch_end {
+                    Some(next_epoch_end) => next_epoch_end,
+                    None => {
+                        let mut state = self.state_at_slot(
+                            epoch.safe_add(1)?.end_slot(T::EthSpec::slots_per_epoch()),
+                            StateSkipConfig::WithoutStateRoots,
+                        )?;
+                        state.build_committee_cache(RelativeEpoch::Current, &self.spec)?;
+                        next_epoch_end.get_or_insert(state)
+                    }
+                };
+
+                // If the next epoch end is no longer phase0, no proposer rewards are awarded, as Altair epoch boundry
+                // processing kicks in. We check this here, as we know that current_epoch_end will always be phase0.
+                if !matches!(next_epoch_end, BeaconState::Base(_)) {
+                    continue;
+                }
+
+                next_epoch_end
+            } else if attestation.data().target.epoch == epoch.safe_sub(1)? {
+                match &mut current_epoch_end {
+                    Some(current_epoch_end) => current_epoch_end,
+                    None => {
+                        let mut state = self.state_at_slot(
+                            epoch.end_slot(T::EthSpec::slots_per_epoch()),
+                            StateSkipConfig::WithoutStateRoots,
+                        )?;
+                        state.build_committee_cache(RelativeEpoch::Current, &self.spec)?;
+                        current_epoch_end.get_or_insert(state)
+                    }
+                }
+            } else {
+                return Err(BeaconChainError::BlockRewardAttestationError);
+            };
+
+            let sqrt_total_active_balance =
+                SqrtTotalActiveBalance::new(processing_epoch_end.get_total_active_balance()?);
+            for attester in get_attesting_indices_from_state(state, attestation)? {
```

Comment:
This algorithm is not quite right because it pays out for attestations regardless of whether this proposer was the the first one to include them.

In the spec, the epoch processing function iterates the `previous_epoch_attestations` and pays the proposer reward to the proposer who includes each attester's attestation with minimal `inclusion_delay`:

```python
attestation = min([
    a for a in matching_source_attestations
    if index in get_attesting_indices(state, a)
], key=lambda a: a.inclusion_delay)
rewards[attestation.proposer_index] += get_proposer_reward(state, index)
```

It's quite common that a validator's attestation will be included multiple times in different aggregates, e.g. first at slot 10 and then again at slot 11. Only the proposer of the block of slot 10 gets the reward in this case (the slot 11 proposer either gets nothing for the attestation if it covered no new validators, or _just_ the rewards for the newly covered validators for which this is their first inclusion on chain).

I think to remedy this the simplest fix would be to:

- For each attester in each attestation included: check whether this is the inclusion with minimal `inclusion_delay` (usually the _first_ inclusion, but not always). This would involve checking first against `state.previous_epoch_attestations`/`state.current_epoch_attestations` as appropriate and then checking against attestations already processed in this block (we don't want to double pay for multi inclusions).

One complication is the `min` by `inclusion_delay`. It would be helpful if a property like this held:

> If an attestation is included for validator `V` at slot `N` in epoch `N // 32` then it has a lower `inclusion_delay` than any other attestation from `V` included at a slot `M > N`.

Unfortunately this property does not hold in the case of slashable attestations made by validator `V` when the chain is not finalising promptly, in which case it could be that:

- There are 2+ chains with different shufflings that descend from the same `source` checkpoint (this requires finality to lapse for >2 epochs as shufflings only diverge if their common ancestor block is >2 epochs ago).
- Validator `V` signs two slashable attestations in epoch `N // 32`: one on each chain. Let `M = N + 1` for simplicity. On the first chain they attest at e.g. slot `N` and on the second chain they attest at slot `N - 5`. The attestation from slot `N - 5` is included at `N` for an `inclusion_delay` of 5, then the attestation from slot `N` is included at slot `M = N + 1` for an `inclusion_delay` of 1. So it's possible for an attestation included at a later block to have a lower `inclusion_delay`, i.e. the property is false.

Therefore in the general case it's not safe to just use the state at the block's slot (`state`) to infer the rewards. I had hoped that if this property held we could use it to avoid loading the `current_epoch_end`/`next_epoch_end` states.

Instead I think we should just proceed with loading all 3 states most of the time and rely on the fact that all the phase0 states are quite small and should load quickly with the soon-to-be-merged hierarchical state diffs:

- https://github.com/sigp/lighthouse/pull/5978

Reply:
Thanks for the review! 
I've implemented a potential fix and a (rather messy) test for this.

I have a question about the second part: While such a pair of attestations would be obviously slashable, would they be really be includable into the "other" chain, respectively? As the shuffling is different there, the attesting validator would likely be in a different committee for a different slot in each chain and therefore the validation would fail (for one of the attestations), as the signature does not match the expected validator, as per my understanding.

Reply:
The beacon node is able to verify attestations from multiple chains, it will load the relevant state & committees and use those to verify the signature. Only in the case where it views the blocks from the other chain as completely invalid will it fail to process the blocks & attestations from that chain.

Reply:
Oh no actually you're right. We can't include the attestations from the other fork on chain once the shufflings diverge, because `process_attestation` in the spec uses the committees from the current chain 

I think this means the property I stated _does_ hold. Which means we could use the passed in `state` for all of the calculations :thinking:

Reply:
While I currently also think that your statement holds, I don't see why we can avoid the state loads.

In phase0, attestation rewards are processed at the epoch boundary. Therefore, we need the state, as a validator might be slashed at the epoch boundary but not at slot `N` where the attestation is included. Furthermore, we need the effective balance and total effective balance from the epoch boundary to properly calculate rewards.

Reply:
Oooh yes, the slashing thing definitely prevents us from using the pre-state. Good catch.

The effective balances are immutable within an epoch, but you're right they'd also be a problem for the current epoch attestations we include which don't get rewarded until the end of the next epoch. Let's leave it as-is :sweat_smile: